{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<center>\n",
    "<h1>Workshop 2: Quality Control (QC)</h1><br><br>\n",
    "<i><big>Preparing and filtering data for genome-wide association studies.</big><br><br>\n",
    "Hatzikotoulas Konstantinos (Kostas) (hatzikotoulas@helmholtz-muenchen.de)</i>\n",
    "</center>\n",
    "\n",
    "## Objectives\n",
    "\n",
    "In this workshop, you will learn the data quality assessment and control steps that are typically carried out during genome wide association studies (GWAS).\n",
    "\n",
    "## Why do we need the QC?\n",
    "\n",
    "Study design and errors in genotype calling can introduce systematic biases into GWAS, leading to spurious associations.\n",
    "A thorough QC can help us identify samples and markers that should be removed prior to association analysis in order to minimize the number of false-positive and false-negative associations. \n",
    "\n",
    "<img src=\"FN.png\" width=\"40%\">\n",
    "\n",
    "In this practical, we assume that the study design has been conducted appropriately and the QC applies to genotypes after they have been called from probe intensity data.\n",
    "\n",
    "## QC protocol\n",
    "\n",
    "The QC protocol of a GWAS is usually split into two broad categories, “Sample QC” and “Variant QC”.\n",
    "Sample QC is done prior to Variant QC because we want to maximise the number of markers remaining in the study.\n",
    "\n",
    "We will be using <code>plink</code> to run the QC and R to visualise the results.\n",
    "You can find a manual and command reference of <code>plink</code> [here](https://www.cog-genomics.org/plink/1.9/) and [here](https://www.cog-genomics.org/plink/1.9/index).\n",
    "\n",
    "## Sample QC\n",
    "\n",
    "It consists of (at least) five steps:\n",
    "\n",
    "<img src=\"QCsteps.png\" width=\"40%\">\n",
    "\n",
    "\n",
    "\n",
    "<br>\n",
    "Initial setup:\n",
    "<br>\n",
    "\n",
    "<br>\n",
    "\n",
    "<div class=\"alert alert-warning\">\n",
    "<b>Location:</b> The data for this workshop can be found in this directory: <code>/dss/dsshome1/lxc06/gobi012/GWAS/gobi/data/Workshop2_QC/</code>. \n",
    "\n",
    "</div>\n",
    "\n",
    "...but for this workshop we have created a container with all files/software needed\n",
    "\n",
    "\n",
    "<br>\n",
    "Initial setup:\n",
    "<br>\n",
    "\n",
    "```bash\n",
    "# login\n",
    "module load charliecloud\n",
    "ch-run /dss/dsshome1/lxc06/gobi012/GWAS/gobi -- ls -lrt --color=auto /data\n",
    "ch-run /dss/dsshome1/lxc06/gobi012/GWAS/gobi -- /bin/bash\n",
    "## test plink runs in the container:\n",
    "plink --help\n",
    "## test that R runs for you, then use q() and then n to exit R\n",
    "R\n",
    "\n",
    "#Once in the container move to your home directory \n",
    "cd ~\n",
    "\n",
    "#Make your own directory and give the path of your DIR and initial unqced FILE\n",
    "mkdir GWASgobi[group number]_QC_[your name]\n",
    "cd GWASgobi[group number]_QC_[your name]\n",
    "(eg mkdir GWASgobi012_QC_Kostas)\n",
    "\n",
    "#Copy the initial UnQCed data to your present working directory (pwd)\n",
    "\n",
    "scp -pr /data/Workshop2_QC/* ./\n",
    "\n",
    "DIR=<path to your directory>\n",
    "\n",
    "(eg DIR=/home/gobi012/GWASgobi012_QC_Kostas)\n",
    "\n",
    "FILE=VSS\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "For each quesiton please provide the following in a text file or as markdown in the Jupyter note book:\n",
    "- Answers\n",
    "- Any code generated\n",
    "For any plots please export them as .png, .jpg or .pdf and label them as GWAS.[question number].\n",
    "\n",
    "Each question answered correctly will be awarded 0.89 point. 25 points in total for practical 2\n",
    "\n",
    "\n",
    "#### In the 5 first steps we will generate all of the files required by using plink and then we will visualise the results in R.\n",
    "\n",
    "### Step_1: Individuals with outlying missing genotype\n",
    "\n",
    "#### Call rate (CR)\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 1a:</b> Run missingness across file genome-wide </div>\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 1a:</b> \n",
    "\n",
    "``` \n",
    "plink --bfile $DIR/$FILE --missing --out $DIR/$FILE-missing \n",
    "(this is an example of what we expect to see as an answer)\n",
    "\n",
    "```\n",
    "</div> \n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 1b:</b> Use plink to exclude samples with a call rate less than 0.98 (use VSS.bed/bim/fam) </div>\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE --mind 0.02 --make-bed --out $VSS-mind\n",
    "```\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 1b:</b> </div>\n",
    "\n",
    "### Step_2: Individuals with discordant sex information\n",
    "\n",
    "#### Sex check\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 2a:</b> Run sex checking (use VSS.bed/bim/fam) </div>\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 2a:</b> </div>\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE --check-sex --make-bed --out $FILE-checksex\n",
    "```\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 2b:</b> Extract Xchr SNPs (use VSS.bed/bim/fam) </div>\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 2b:</b> </div>\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE --chr X --make-bed --out $FILE-xchrsnps\n",
    "```\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 2c:</b> Run missingness on Xchr SNPs </div>\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 2c:</b> </div>\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-xchrsnps --missing --make-bed --out $FILE-xchrsnps-missing\n",
    "```\n",
    "\n",
    "### Step_3: Individuals with outlying heterozygosity rate\n",
    "\n",
    "#### Heterozygosity\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 3a:</b> Produce a file containing only the autosomal variants (use VSS.bed/bim/fam) </div>\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 3a:</b> </div>\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE --autosome --make-bed --out $FILE-autosomal\n",
    "```\n",
    "\n",
    "We need to split the dataset into rare/low frequency and common minor allele frequency (MAF) and look at the missingness separately for each\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 3b:</b> Extract autosomal SNPs with MAF greater than/equal to 1% </div>\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 3b:</b> </div>\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-autosomal --maf 0.01 --het --make-bed -out $FILE-mafgre1p\n",
    "```\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 3c:</b> Extract autosomal SNPs with MAF less than 1% </div>\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 3c:</b> </div>\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-autosomal --max-maf 0.01 --het --make-bed --out $FILE-mafles1p\n",
    "```\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 3d:</b> Get missingness of each MAF file (eg greater than/equal to 1% and less than 1%) to plot against heterozygosity in R </div>\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-mafles1p --missing --make-bed --out $FILE-mafles1p-miss\n",
    "\n",
    "```\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-mafgre1p --missing --make-bed -out $FILE-mafgre1p-miss\n",
    "```\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 3d:</b> </div>\n",
    "\n",
    "### Step_4: Duplicated or related individuals\n",
    "\n",
    "#### Relatedness/Duplicates\n",
    "\n",
    "Pair-wise IBD to look at duplicates.A) Using only variants ≥1%, excluding complex regions and B) LD prune using R-squared 0.2, a window size of 50 variants and 5 variants to shift the window at the end of each step\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 4a:</b> Use autosomal SNPs with MAF ≥1% and exclude complex regions (please use this file: complex_regions.txt)</div>\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 4a:</b> </div>\n",
    "\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-mafgre1p --exclude complex_regions.txt --make-bed --out $FILE-excludecomplex\n",
    "```\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 4b:</b> LD prune </div>\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 4b:</b> </div>\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-mafgre1p --indep-pairwise 50['kb'] 5 0.2 --make-bed --out $FILE-LDprune\n",
    "```\n",
    "\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 4c:</b> Calculate pair-wise IBD </div>\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 4c:</b> </div>\n",
    "\n",
    "```\n",
    "plink --bfile $DIR/$FILE-mafgre1p --genome --make-bed --out $FILE-ibd\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "### Step_5: Ethnicity outliers\n",
    "\n",
    "#### Ethnicity MDS distance matrix\n",
    "\n",
    "For this step, you will need to merge your data with 1000 Genomes genotype data (or HapMap genotype data).\n",
    "(Its a time consuming procedure thus please use the following file)\n",
    "\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "FILE2=VSS-1Kg\n",
    "\n",
    "```\n",
    "\n",
    "##### Pair-wise IBD.\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 5a:</b> Use autosomal SNPs with MAF ≥1% and exclude complex regions. </div>\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 5a:</b> </div>\n",
    "\n",
    "\n",
    "``\n",
    "plink --bfile $DIR/$FILE2 --maf --exclude complex_regions.txt --make-bed --out $FILE2-gre1p-exclucomplex\n",
    "``\n",
    "\n",
    "\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 5b:</b> LD prune using R-squared 0.2, a window size of 50 variants and 5 variants to shift the window at the end of each step. </div>\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 5b:</b> </div>\n",
    "\n",
    "``\n",
    "plink --bfile $DIR/$FILE2-gre1p-exclucomplex --indep-pairwise 50['kb'] 5 0.2 --make-bed --out $FILE2-LDprune\n",
    "``\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 5c:</b> Calclulate pair-wise IBD </div>\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 5c:</b> </div>\n",
    "\n",
    "``\n",
    "plink --bfile $DIR/$FILE2-LDprune --genome --make-bed --out $FILE2-ibd\n",
    "``\n",
    "\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 5d:</b> MDS distance matrix calcualting the first 10 components </div>\n",
    "\n",
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 5d:</b> </div>\n",
    "\n",
    "``\n",
    "plink --bfile  $DIR/$FILE2-ibd --pca 10 --make-bed --out $FILE2-pca10\n",
    "``"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now, we need to check our results against R plots\n",
    "\n",
    "%%R\n",
    "\n",
    "##### -----------------------------------------------------------------------------------------------------------------#\n",
    "#####         (6)  SAMPLE CALL RATE    - threshold = 98%                                                               #\n",
    "##### -----------------------------------------------------------------------------------------------------------------#\n",
    "\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 6a:</b> Plot the missingness of the individuals (eg a histogram of xlab=missingness and ylab=frequency) </div>\n",
    "<div class=\"alert alert-success\"><b>Question 6b:</b> Create a .txt file with the samples to be removed (missingness of >2%) </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "missingData <- read.table(VSS-missing.imiss)\n",
    "hist(missingData$F_MISS, xlab = 'missingness', ylab='frequency', main = \"Missingness of Individuals\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "removedData <- read.table(VSS-mind.irem)\n",
    "write.table(removedData, file = \"VSS-mind.txt\", sep = \"\\t\", row.names = FALSE, col.names = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### -----------------------------------------------------------------------------------------------------------------#\n",
    "#####          (7) SEX CHECKING                                                                                         #\n",
    "##### -----------------------------------------------------------------------------------------------------------------#\n",
    "<div class=\"alert alert-success\"><b>Question 7a:</b> Plot the sex-check results (eg xlab=\"X chr inbreeding (homozygosity) estimate F\", ylab=\"Proportion of missing SNPs for the X chr\") </div>\n",
    "<div class=\"alert alert-success\"><b>Question 7b:</b> Create a .txt file with the samples to be removed (status:problem) </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "checksexData <- read.table(\"VSS-checksex.sexcheck\", h=T)\n",
    "xchrsnpsMissData <- read.table(\"VSS-xchrsnps-missing.imiss\",h=T)\n",
    "\n",
    "sex_xchrmData <- merge(checksexData,xchrsnpsMissData, by=c(\"FID\",\"IID\"))\n",
    "plot(sex_xchrmData$F,sex_xchrmData$N_MISS,xlab=\"X chr inbreeding (homozygosity) estimate F\", \n",
    "     ylab=\"Proportion of missing SNPs for the X chr\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sex-problemData <- subset(sex_xchrmData, STATUS = \"PROBLEM\", select = IID)\n",
    "write.table(sex_problemData, file = \"sex-problem.txt\",sep =\"\\t\", row.names = FALSE, col.names = FALSE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### -----------------------------------------------------------------------------------------------------------------#\n",
    "#####      (8)    Heterozygosity   MAF≥1% and MAF<1%                                                                   #\n",
    "##### -----------------------------------------------------------------------------------------------------------------#\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 8a:</b> Plot the heterozygosity results (maf>=1%) against missingness (eg xlab=\"% Heterozygosity\", ylab=\"F_MISS\") </div>\n",
    "<div class=\"alert alert-success\"><b>Question 8b:</b> Create a .txt file with the samples to be removed (more than 3xSD away from the mean) </div>\n",
    "<div class=\"alert alert-success\"><b>Question 8c:</b> Repeat the above steps for the heterozygosity results of maf less than 1% </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'mafg1' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-1-88bc4a04ab60>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mmafg1\u001b[0m\u001b[1;33m-\u001b[0m \u001b[0mread\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtable\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mVSS\u001b[0m\u001b[1;33m-\u001b[0m\u001b[0mgre1p\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'mafg1' is not defined"
     ]
    }
   ],
   "source": [
    "mafg1miss <- read.table(\"VSS-mafgre1p-miss.imiss\", h =T)\n",
    "\n",
    "\n",
    "mafg1het <- read.table(\"VSS-mafgre1p.het\", h = T)\n",
    "\n",
    "g1hemiss <- merge(mafg1miss, mafg1het, by = c(\"FID\",\"IID\"))\n",
    "\n",
    "\n",
    "names(g1hemiss)[7] <- \"O\"\n",
    "names(g1hemiss)[9] <- \"N\"\n",
    "\n",
    "g1hemiss$O  <- as.numeric(g1hemiss$O)\n",
    "g1hemiss$N  <- as.numeric(g1hemiss$N)\n",
    "\n",
    "plot(g1hemiss$O/g1hemiss$N, g1hemiss$F_MISS,xlab=\"% Heterozygosity\", ylab=\"F_MISS\", main =\"Heterozygosity MAF≥1%\" )\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mafles1miss <- read.table(\"VSS-mafles1p-miss.imiss\",h= T)\n",
    "\n",
    "mafles1het <- read.table(\"VSS-mafles1p.het\", h=T)\n",
    "\n",
    "les1hemiss <- merge(mafles1miss, mafles1het, by = c(\"FID\",\"IID\"))\n",
    "\n",
    "\n",
    "names(les1hemiss)[7] <- \"O\"\n",
    "names(les1hemiss)[9] <- \"N\"\n",
    "\n",
    "les1hemiss$O  <- as.numeric(les1hemiss$O)\n",
    "les1hemiss$N  <- as.numeric(les1hemiss$N)\n",
    "\n",
    "plot(les1hemiss$O/les1hemiss$N, les1hemiss$F_MISS,xlab=\"% Heterozygosity\", ylab=\"F_MISS\", main = \"Heterozygosity MAF<1%\" )\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### -----------------------------------------------------------------------------------------------------------------#\n",
    "#####      (9)   DUPLICATES - RELATEDNESS                                  \n",
    "##### -----------------------------------------------------------------------------------------------------------------#\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 9a:</b> Plot the IBD results (eg PI_HAT) </div>\n",
    "<div class=\"alert alert-success\"><b>Question 9b:</b> Create 2 .txt files with the samples to be removed. One for duplicates/twins and one for related individuals of a PI_HAT value >0.2 </div>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ibd_result <- read.table(\"VSS-ibd.genome\",h = T)\n",
    "hist(ibd_result$PI_HAT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "identicalData <- subset(ibd_result, PI_HAT > 0.9, select = c(IID1,IID2,Z2))\n",
    "write.table(identicalData, file = \"VSS-identical.txt\", sep = \"\\t\", row.names = FALSE, col.names = TRUE)\n",
    "\n",
    "relativData <-  subset(ibd_result, PI_HAT > 0.2, select = c(IID1,IID2,PI_HAT))\n",
    "write.table(relativData, file = \"VSS-relative.txt\", sep = \"\\t\", row.names = FALSE, col.names = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### -----------------------------------------------------------------------------------------------------------------#\n",
    "#####      (10)   ETHNICITY           Merged with 1000 genomes file                                                     #\n",
    "##### -----------------------------------------------------------------------------------------------------------------#\n",
    "\n",
    "#### The Populationfile-1kg.txt contains individuals' ethnicities\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 10a:</b> Merge the population information (Populationfile-1kg.txt) into the genome data and plot PC1 and PC2 - subsetting by population (eg use different coulours) </div>\n",
    "<div class=\"alert alert-success\"><b>Question 10b:</b> Identify the outliers and create a .txt file  </div>\n",
    "<div class=\"alert alert-success\"><b>Question 10c:</b> Remove the outliers from the population file and plot again </div>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "population_data <- read.table(\"Populationfile-1kg.txt\",h =T)\n",
    "vss1kgibd <- read.table(\"VSS-1Kg-pca10.eigenvet\")\n",
    "\n",
    "names(vss1kgpca)[2] <- \"ID\"\n",
    "\n",
    "pca_pop_data <- merge(population_data,vss1kgpca, by(ID))\n",
    "\n",
    "plot(pca_pop_data$V3,pca_pop_data$V4, col = \"Population\")\n",
    "\n",
    "\n",
    "\n",
    "ggplot(pca_pop_data,aes(V3,V4, color = Population))+geom_point()+\n",
    "labs(title =\"PC1 and PC2 - subsetting by population\", x = \"PC1\", y = \"PC2\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "population_data <- read.table(\"Populationfile-1kg.txt\",h =T)\n",
    "vss1kgpca <- read.table(\"VSS-1Kg-pca10.eigenvec\")\n",
    "\n",
    "names(vss1kgpca)[2] <- \"ID\"\n",
    "\n",
    "pca_pop_data <- merge(population_data,vss1kgpca, by = \"ID\")\n",
    "\n",
    "U <- pca_pop_data[\"V3\"]\n",
    "U2 <- pca_pop_data[\"V4\"]\n",
    "\n",
    "\n",
    "apply(U, 2, function(x) which( abs(x - mean(x)) > (6 * sd(x)) ))\n",
    "apply(U2, 2, function(x) which( abs(x - mean(x)) > (6 * sd(x)) ))\n",
    "pca_pop_data2 <- data.frame(pca_pop_data)\n",
    "\n",
    "U_id <- subset(pca_pop_data2, !((V3 %in% U)&(V4 %in% U2)), select = c(ID,Population,Sex, V3,V4))\n",
    "\n",
    "new_data_pca <- subset(pca_pop_data2, !(V3 %in% U))\n",
    "new_data_pca <- subset(new_data_pca, !(V4 %in% U2))\n",
    "\n",
    "\n",
    "write.table(U_id, file = \"PCA_outliers.txt\", sep = \"\\t\", row.names = FALSE, col.names = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "ggplot(pca_pop_data2,aes(V3,V4, color = Population))+\n",
    "geom_point()+\n",
    "labs(title =\"PC1 and PC2 - subsetting by population\", x = \"PC1\", y = \"PC2\")\n",
    "\n",
    "ggplot(U_id,aes(V3,V4, color = Population))+\n",
    "geom_point()+\n",
    "labs(title =\"PC1 and PC2 - subsetting by population\", x = \"PC1\", y = \"PC2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Combine all outliers' files and exclude them from your data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variant QC\n",
    "\n",
    "It consists of (at least) four steps:\n",
    "\n",
    "    1.Identification of variants with an excessive missing genotype\n",
    "    2.Identification of variants demonstrating a significant deviation from Hardy-Weinberg equilibrium (HWE)\n",
    "    3.Removal of all makers with a very low minor allele frequency\n",
    "    4.Removal of all makers with cluster separation score <0.4\n",
    "\n",
    "The criteria used to filter out low quality markers differ from study to study.\n",
    "Variant QC must be done with great care as every removed marker is potentially a missed disease variant.\n",
    "Imputation can be used to recover some of the excluded markers\n",
    "\n",
    "Here we are using the following thresholds:\n",
    "\n",
    "- Call rate = 98%\n",
    "- HWE       = p <=1x10-4\n",
    "- MAF       = 1%\n",
    "- Cluster separation score = <=0.4\n",
    "\n",
    "<div class=\"alert alert-success\"><b>Question 11a:</b> Run the plink command for all these thresholds. The cluster separation scores are provided to you (coreex_argogs_20141015.clustersep.txt) </div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\" role=\"alert\"><b>Answer 11a:</b> </div>\n",
    "\n",
    "``\n",
    "plink --bfile  $DIR/$FILE --geno 0.02 --hwe 1e-4 \"midp\" --maf 0.01 --exclude coreex_argogs_20141015.clustersep.txt --make-bed --out $FILE-variantqc\n",
    "``"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
